AzureRmKubernetes cmdlets for managed Kubernetes

devigned opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers the changes made by the pull request.

If applicable, reference the bug/issue that this pull request fixes here.
-->

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

twitchax commented:
Restarted the travis build: no idea why it died.

twitchax commented:
Looks like there are some errors with api version namespaces?

maddieclayton commented:
@devigned I'll assign this back to you for the moment, please reassign it to me when you're ready for me to take a look. Thanks!

markcowl commented on src/ResourceManager/Kubernetes/AzureRM.Kubernetes.Netcore.psd1:
```diff
@@ -0,0 +1,125 @@
+    PSData = @{
+
+        # Tags applied to this module. These help with module discovery in online galleries.
+        Tags = 'Azure','Kubernetes','ARM','Provider','AKS'
```
Needs the tags that indicates this runs on core (Linux, Mac) - this is the mechanism that the posh team is using to differentiate module compatible with PowerShell Core.

devigned replied:
What tags would those be? As far as I can tell, AzureRM.Profile.Netcore.psd1 doesn't have any tags that jump out at me which scream netcore (`Tags = 'Azure','ResourceManager','ARM','Profile','Authentication','Environment','Subscription'`).

markcowl commented on src/ResourceManager/Kubernetes/AzureRM.Kubernetes.psd1:
```diff
@@ -0,0 +1,124 @@
+RequiredModules = @(@{ModuleName = 'AzureRM.Profile'; ModuleVersion = '4.0.0'; })
+
+# Assemblies that must be loaded prior to importing this module
+RequiredAssemblies = '.\YamlDotNet.dll'
```
All assemblies required by the cmdlet assembly should go here, normally this includes the associated management library.

devigned replied:
The generated management library is included in the command module. This will need to be removed and replaced with a reference when the SDK is produced.

markcowl commented on src/ResourceManager/Kubernetes/AzureRM.Kubernetes.psd1:
```diff
@@ -0,0 +1,124 @@
+    PSData = @{
+
+        # Tags applied to this module. These help with module discovery in online galleries.
+        Tags = 'Azure','ResourceManager','ARM','AKS', 'Kubernetes'
```
Container?

devigned replied:
Sure!

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes.Test.Netcore/MergeKubeConfigTests.cs:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        [Fact]
+        public void MergedKubeConfigOverwriteClustersTest()
```
Use traits to mark these as check-in tests (```[Trait(Category.AcceptanceType, Category.CheckIn)]``` these constants are in the common scenario test assembly

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes.Test.Netcore/ScenarioTests/KubernetesTests.ps1:
```diff
@@ -0,0 +1,22 @@
+    $location = Get-ProviderLocation "Microsoft.ContainerService/managedClusters"
+
+    New-AzureRmResourceGroup -Name $resourceGroupName -Location $location
+
```
I am assuming this will contain a CRUD test at some point.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes.Test/MergeKubeConfigTests.cs:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        [Fact]
+        public void MergedKubeConfigAdditiveContextTest()
```
same comments on traits

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands.Kubernetes.csproj:
```diff
@@ -0,0 +1,151 @@
+      <HintPath>..\..\..\packages\AutoMapper.6.2.2\lib\net45\AutoMapper.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="System.Management.Automation, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
You should not need to get this from Nuget

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands.Kubernetes.csproj:
```diff
@@ -0,0 +1,151 @@
+      <Project>{70527617-7598-4aef-b5bd-db9186b8184b}</Project>
+      <Name>Commands.Common.Authentication.Abstractions</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\..\..\Common\Commands.Common.Authorization\Commands.Common.Authorization.csproj">
```
SHould not take this reference - Commands.Common.Authentication is meant to contain realizations of the Abstractions types, no cmdlet assembly other than profile should take a dependency on thsi assembly

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet(VerbsCommon.New, KubeNounStr, DefaultParameterSetName = DefaultParamSet)]
+    [OutputType(typeof(PSObject), typeof(List<PSObject>))]
```
We should avoid using PSObject as the type - we should be using strongly types parameters and output types.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage = "The secret associated with the AAD application / service principal.")]
+        public string ClientSecret { get; set; }
```
This should be a secure string, or ClientId + ClientSecret could be represented as a PSCredetial

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+
+
+        [Parameter(Mandatory = false, HelpMessage = "The version of Kubernetes to use for creating the cluster.")]
+        [ValidateSet("1.7.7", "1.8.1")]
```
I am assuming that this will change over time - if so, this shoudl use a GenericArgumentCompleter, rather than a ValidateSet

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+        public int NodeCount { get; set; } = 3;
+
+        [Parameter(Mandatory = false, HelpMessage = "The default number of nodes for the node pools.")]
+        public int? NodeOsDiskSize { get; set; }
```
Why nullable?  The correct pattern is to use ```MyInvocation.BoundParameters``` to determine if a property is set, rather than relying on the cmdlet  property being set to null.

devigned replied:
Oops! It is actually used that way in the Set cmdlet. I will remove the nullable.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "SSH key file value or key file path. Defaults to {HOME}/.ssh/id_rsa.pub.")]
+        public string SshKeyValue { get; set; }
```
Include a parameter alias for SSHKeyPath, so it is clear thsi can be a path.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+        protected string GetSshKey(string sshKeyOrFile)
+        {
+            // SSH key was specified as either a file or as key data
+            if (!string.IsNullOrEmpty(SshKeyValue))
```
need braces to define the boundaries of the If block

devigned replied:
Done.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+            AddRoleAssignment("Contributor", app.AppId);
+        }
+
+        protected bool Exists()
```
More descriptive function name

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+                catch (Exception)
+                {
+                    // AAD might puke here, so we catch it and try again until success
+                    Thread.Sleep(1000 * i);
```
Use [TestMockSupport.Delay](https://github.com/Azure/azure-powershell/blob/preview/src/Common/Commands.Common/TestMockSupport.cs#L34) instead of calling Thread.Sleep directly

devigned replied:
This is not a test. I repeat this is not a test. 

No, but seriously, this needs to be here because sometimes we need to wait a bit for AAD to propagate changes so that we can add the ARM RBAC role to the service principal.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,363 @@
+            File.WriteAllText(AcsSpFilePath, JsonConvert.SerializeObject(config));
+        }
+
+        protected static string RandomUtfString(int size)
```
Many of these seems like utility functions that should be factored into a utility class, or implementd as extensions over the base cmdlet type.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/ImportAzureRmKubernetesCredential.cs:
```diff
@@ -0,0 +1,247 @@
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet("Import", KubeNounStr + "Credential", SupportsShouldProcess = true, DefaultParameterSetName = GroupNameParameterSet)]
+    [OutputType(typeof(string))]
```
use Void if there is no output

devigned replied:
It is string or pass thru. If someone uses -ConfigPath "-", then it will output the config file as a string, else pass thru if -PassThru.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/NewAzureRmKubernetes.cs:
```diff
@@ -0,0 +1,64 @@
+{
+    [Cmdlet(VerbsCommon.New, KubeNounStr, DefaultParameterSetName = DefaultParamSet, SupportsShouldProcess = true)]
+    [OutputType(typeof(PSKubernetesCluster))]
+    public class New : CreateOrUpdateKubeBase
```
Name the cmdlet class as the cmdlet name (without dashes)

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/NewAzureRmKubernetes.cs:
```diff
@@ -0,0 +1,64 @@
+    {
+        [Parameter(Mandatory = false, HelpMessage = "Create cluster even if it already exists")]
+        public SwitchParameter Force { get; set; }
+
```
Should have an AsJob parameter, which should get you Jobs for free.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/RemoveAzureRmKubernetes.cs:
```diff
@@ -0,0 +1,112 @@
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet(VerbsCommon.Remove, KubeNounStr, SupportsShouldProcess = true, DefaultParameterSetName = GroupNameParameterSet)]
+    public class Remove : KubeCmdletBase
```
same comment

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/RemoveAzureRmKubernetes.cs:
```diff
@@ -0,0 +1,112 @@
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(Mandatory = false)]
+        public SwitchParameter PassThru { get; set; }
```
Also should have an AsJob parameter

devigned replied:
Added in https://github.com/Azure/azure-powershell/pull/5087/commits/57ae5f27b10712185b74ec22325a89e469512188

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,330 @@
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet("Start", KubeNounStr + "Dashboard")]
+    public class StartDashboard : KubeCmdletBase
```
OutputType?

devigned replied:
I'll change it to typeof(KubeTunnelJob).

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,330 @@
+
+                var proxyUrl = "http://127.0.0.1:8001";
+
+                WriteVerbose(string.Format(
```
We should likely look for kubectl on the path and write a nice error if it does not exist before oing further.

devigned replied:
Done. See GeneralUtilities.Probe for implementation.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,330 @@
+
+                    if (PassThru)
+                    {
+                        WriteObject(true);
```
You should return the job

devigned replied:
will do

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,330 @@
+            _dashPod = dashPod;
+        }
+
+        public override string StatusMessage { get; }
```
need to implement.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StopAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,51 @@
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet("Stop", KubeNounStr + "Dashboard")]
+    public class StopDashboard : KubeCmdletBase
```
I don't know that we need this.  If we return a job from the first cmdlet, the user can simply stop it.

devigned replied:
It not needed by someone that understands powershell jobs, but it does give a simple way to kill the job if you are not adept with Jobs.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/load.ps1:
```diff
@@ -0,0 +1,3 @@
+Import-Module -Name ./src/Package/Release/ResourceManager/AzureRM.Profile.Netcore/AzureRM.Profile.Netcore.psd1
```
??

devigned replied:
This was just for testing. I'll drop it before the PR is final.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Stop-AzureRmKubernetesDashboard.md:
````diff
@@ -0,0 +1,53 @@
+PS C:\> Stop-AzureRmKubernetes
+```
+
+{{ Add example description here }}
````
Please ensure that you fill these out in the markdown files.

markcowl commented on src/ResourceManager/Kubernetes/NuGet.Config:
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
```
I think the changes below this are unintentional, right?

maddieclayton commented:
@devigned Looks like you are failing because of some signature issues.  If you check out the SignatureIssues.csv [here](https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/1237/artifact/src/Package/) you can see what is causing the failure.  You can ignore severity 2 issues for the most part, but you should address the sev 1 issues.

maddieclayton commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,361 @@
+
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet(VerbsCommon.New, KubeNounStr, DefaultParameterSetName = DefaultParamSet)]
```
All action cmdlets should include supportsshouldprocess: more info [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess)

maddieclayton commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,361 @@
+        public SwitchParameter AsJob { get; set; }
+
+        [Parameter(Mandatory = false)]
+        public Hashtable Tags { get; set; }
```
Rename to "Tag" (more info [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#singular-noun-1))

devigned commented:
@markcowl and @maddieclayton I the PR has had most of the boulders excavated and is ready for nits and finer review. I believe I've addressed all of the feedback thus far. I think I need to clean up some signing errors, but that should be about it. Let me know if there are any other changes you'd like me to make.

maddieclayton commented:
@azuresdkci Test this please

maddieclayton commented:
@devigned That's a Batch test that fails intermittently, started the build again.

devigned commented:
@markcowl or @maddieclayton any other issue, or can this be merged?

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,346 @@
+            switch (ParameterSetName)
+            {
+                case IdParameterSet:
+                    {
```
no reason for braces here

devigned replied:
It is needed due to 'var resource = ...' in both blocks. This limits scope.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,346 @@
+            RunCmdLet(() =>
+            {
+                if (!GeneralUtilities.Probe("kubectl"))
+                    throw new CmdletInvocationException("kubectl is required to be installed and on your path to execute this command. Kubectl is available here: https://kubernetes.io/docs/tasks/tools/install-kubectl/.");
```
braces around here.  .Net coding guidelines are sometimes weird and arbitrary

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,346 @@
+                var tmpFileName = Path.GetTempFileName();
+                var encoded = Client.ManagedClusters.GetAccessProfiles(ResourceGroupName, Name, "clusterUser")
+                    .KubeConfig;
+                File.WriteAllText(
```
Use IDataStore

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,346 @@
+                var proxyUrl = "http://127.0.0.1:8001";
+
+                WriteVerbose(string.Format(
+                    "Running: kubectl get pods --kubeconfig {0} --namespace kube-system --output name --selector k8s-app=kubernetes-dashboard",
```
required: resource strings, and elsewhere in this file as well

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StartAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,346 @@
+            {
+                _statusMsg = "pid doesn't exit or job is already dead";
+            }
+            SetJobState(JobState.Stopped);
```
vertical space after closing brace

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/New-AzureRmKubernetes.md:
````diff
@@ -0,0 +1,264 @@
+### Create a new managed Kubernetes cluster with default params
+```
+PS C:\> New-AzureRmKubernetes -ResourceGroupName group -Name myCluster
+```
````
Sample output

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Remove-AzureRmKubernetes.md:
````diff
@@ -0,0 +1,171 @@
+```
+
+### -InputObject
+A PSKubernetesCluster object, normally passed through the pipeline.```yaml
````
fix formatting

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Remove-AzureRmKubernetes.md:
````diff
@@ -0,0 +1,171 @@
+```
+
+### -Id
+Id of a managed Kubernetes cluster```yaml
````
fix formatting  code ```yaml must be on its own line

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Import-AzureRmKubernetesCredential.md:
````diff
@@ -0,0 +1,205 @@
+```
+
+### -InputObject
+A PSKubernetesCluster object, normally passed through the pipeline.```yaml
````
fix formatting

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Import-AzureRmKubernetesCredential.md:
````diff
@@ -0,0 +1,205 @@
+```
+
+### -Id
+Id of a managed Kubernetes cluster```yaml
````
fix formatting, ```yaml must be on its own line

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Get-AzureRmKubernetes.md:
````diff
@@ -0,0 +1,124 @@
+```
+
+### -Id
+Id of a managed Kubernetes cluster```yaml
````
fix formatting - ```yaml must be on its own line

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Set-AzureRmKubernetes.md:
````diff
@@ -0,0 +1,303 @@
+
+### Example 1
+```
+PS C:\> Get-AzureRmKubernetes -ResourceGroupName group -Name myCluster | Set-AzureRmKubernetes -NodeCount 5
````
Sample output (if this cmdlet has some

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Set-AzureRmKubernetes.md:
````diff
@@ -0,0 +1,303 @@
+```
+
+### -Tags
+{{Fill Tags Description}}
````
Isn't this -Tag?  Also, please fill in the description

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Start-AzureRmKubernetesDashboard.md:
````diff
@@ -0,0 +1,144 @@
+### Start an SSH tunnel and open a browser to the Kubernetes dashboard
+```
+PS C:\> Start-AzureRmKubernetesDashboard -ResourceGroupName group -Name myCluster
+```
````
Sample output

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Start-AzureRmKubernetesDashboard.md:
````diff
@@ -0,0 +1,144 @@
+```
+
+### -Id
+Id of a managed Kubernetes cluster```yaml
````
fix formatting, as above

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/help/Start-AzureRmKubernetesDashboard.md:
````diff
@@ -0,0 +1,144 @@
+```
+
+### -InputObject
+A PSKubernetesCluster object, normally passed through the pipeline.```yaml
````
fix formatting, as above

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/packages.config:
```diff
@@ -0,0 +1,8 @@
+  <package id="Newtonsoft.Json" version="6.0.8" targetFramework="net452" />
+  <package id="System.Management.Automation_PowerShell_3.0" version="6.3.9600.17400" targetFramework="net452" />
+  <package id="YamlDotNet" version="4.2.3" targetFramework="net452" />
+  <package id="YamlDotNet.Signed" version="4.2.2" targetFramework="net452" />
```
??

markcowl commented on tools/StaticAnalysis/Exceptions/SignatureIssues.csv:
```diff
@@ -1547,4 +1547,10 @@
-"Microsoft.Azure.Commands.MachineLearningCompute.dll","Microsoft.Azure.Commands.MachineLearningCompute.Cmdlets.RemoveAzureRmMlOpCluster","Remove-AzureRmMlOpCluster","1","8410","Parameter IncludeAllResources of cmdlet Remove-AzureRmMlOpCluster does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
\ No newline at end of file
+"Microsoft.Azure.Commands.MachineLearningCompute.dll","Microsoft.Azure.Commands.MachineLearningCompute.Cmdlets.RemoveAzureRmMlOpCluster","Remove-AzureRmMlOpCluster","1","8410","Parameter IncludeAllResources of cmdlet Remove-AzureRmMlOpCluster does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
+"Microsoft.Azure.Commands.Kubernetes.dll","Microsoft.Azure.Commands.Kubernetes.GetAzureRmKubernetes","Get-AzureRmKubernetes","1","8400","Get-AzureRmKubernetes uses the noun 'AzureRmKubernetes', which does not follow the enforced naming convention of using a singular noun for a cmdlet name.","Consider using a singular noun for the cmdlet name."
```
I think we should probably add 'Kebernetes' as one of our exceptions to the plurals rule, but this will do for now

devigned replied:
'Kubernetes' :)

markcowl commented:
@Devigned  issues marked as required must be fixed, issues otherwise marked should be fixed before this is stable

markcowl commented on src/ResourceManager/Kubernetes/AzureRM.Kubernetes.psd1:
```diff
@@ -0,0 +1,128 @@
+
+        # External dependent modules of this module
+        # ExternalModuleDependencies = ''
+
```
Add Prerelease = 'preview'  thsi will make this officially a preview module

devigned commented:
Awesome... This just keeps going.

twitchax commented on src/ResourceManager/Kubernetes/Commands.Kubernetes.Test/MergeKubeConfigTests.cs:
```diff
@@ -0,0 +1,107 @@
+
+        public MergedKubeConfigTests()
+        {
+            var merged = ImportAzureRmKubernetesCredential.MergeKubeConfig(File.ReadAllText("./Fixtures/config.yaml"), File.ReadAllText("./Fixtures/additional_kube_config.yaml"));
```
`TestFixture` stuff is funky sometimes, and some examples are bad.

Here is an example of one that I did: https://github.com/twitchax/DocumentDb.Fluent/blob/master/src/Tests/UnitTests.cs.

twitchax commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands.Kubernetes.csproj:
```diff
@@ -0,0 +1,165 @@
+      <HintPath>..\..\..\packages\AutoMapper.6.2.1\lib\net45\AutoMapper.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
```
@markcowl, I think you mean that it is in the `targets` file, correct?

Even if something is loaded by the common runtime, it's not a bad idea to state your true dependencies.

markcowl replied:
@twitchax it is another place we have to change it if we update the common runtime to newtonsoft.json 10, as we plan to do.  If this assembly has a specific dependency on newtonsoft.json, other than in its management client, then this should be removed.

markcowl commented on src/Common/Commands.Common.Tests/ProbeTests.cs:
```diff
@@ -0,0 +1,58 @@
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void TrueWhenProgramDoesExistTest()
+        {
+            Assert.True(GeneralUtilities.Probe("powershell", " -c 'echo hello world!'"));
```
No, I mean if this is run on netcore, the executable is pwsh, ratehr than powershell, isn't it?

devigned commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs:
```diff
@@ -0,0 +1,370 @@
+                var url = string.Format("http://{0}.{1}.{2}.cloudapp.azure.com", salt, DnsNamePrefix, Location);
+
+                acsServicePrincipal = BuildServicePrincipal(Name, url, clientSecret);
+                WriteVerbose(string.Format(
```
I didn't intend on showing these messages unless verbose.

devigned commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/StopAzureRmKubernetesDashboard.cs:
```diff
@@ -0,0 +1,51 @@
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet("Stop", KubeNounStr + "Dashboard")]
+    public class StopAzureRmKubernetesDashboard : KubeCmdletBase
```
I think it is nice for folks who are new to powershell.

devigned commented:
@markcowl can you point to a good use of mocking of a scenario test using the AzureSession.Instance.DataStore? I've looked around, and it seems that most tests are using to validate a file was created, but I couldn't find one that added a file fixture to the DataStore to setup the test.

devigned commented:
@markcowl I've registered YamlDotNet.Signed 4.2.4 with the OSS process, and it's all green.

devigned commented:
@markcowl I just rebased to the latest in preview to deal with a conflict. I think the PR has covered all the required items.

devigned commented:
@markcowl I just rebased again due to another conflicting change merged.

markcowl commented on src/ResourceManager/Kubernetes/Commands.Kubernetes/Microsoft.Azure.Commands.Kubernetes.Format.ps1xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
no need for this file

markcowl commented:
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/409/

cormacpayne commented:
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/414/

markcowl commented:
new on-demand run here: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/422/
